constraints/memory.air

```diff
@@ -2,13 +2,253 @@
+# │ idx0/1  │ Element index within word [0,3] - binary decomposition of low 2 bits     │
+# │   clk   │ Clock cycle when operation occurred                                      │
+# │ v0-v3   │ Four field elements stored in memory word                                │
+# │ d0/d1   │ Delta tracking columns for monotonicity verification                     │
+# │  d_inv  │ Inverse delta column for consecutive transitions                         │
+# │ f_scw   │ Same context/word address flag for sequential access optimization        │
+# └─────────┴──────────────────────────────────────────────────────────────────────────┘
+#
+# STATUS: Partially implemented (missing bus integration)
 #
 # REFERENCES:
 # - Memory Design: https://0xmiden.github.io/miden-vm/design/chiplets/memory.html
 ##########################################################################################
 
+mod memory
+
+use utils::*;
+
+##########################################################################################
+# MEMORY CHIPLET CONSTRAINTS
+##########################################################################################
+
+# Enforces proper memory initialization when entering memory chiplet
+ev memory_chiplet_constraints_first_row([memory[15]]) {
+    let is_read = memory[0];        # Read(1)/Write(0) selector
+    let is_word_access = memory[1]; # Element(0)/Word(1) access selector
+    let ctx = memory[2];            # Context ID
+    let addr = memory[3];           # Memory address
+    let idx0 = memory[4];           # Element index bit 0
+    let idx1 = memory[5];           # Element index bit 1
+    let v0 = memory[7];             # Memory value 0
+    let v1 = memory[8];             # Memory value 1
+    let v2 = memory[9];             # Memory value 2
+    let v3 = memory[10];            # Memory value 3
+
+    # Element selection flags: f[i] = 1 when element index (2*idx1 + idx0) equals i
+    let f0 = binary_not(idx1') * binary_not(idx0');
+    let f1 = binary_not(idx1') * idx0';
+    let f2 = idx1' * binary_not(idx0');
+    let f3 = idx1' * idx0';
+
+    # c_i is set to 1 when `v'[i]` is not written to, and 0 otherwise.
+    #
+    # In other words, c_i is set to 1 when `v'[i]` needs to be constrained (to either 0 or `v[i]`).
+    #
+    # Note that `c_i` only uses values in the "next" row. This is because it must be used to
+    # constrain the first row of the memory chiplet, where that row sits in the "next" position of
+    # the frame, and the "current" row belongs to the previous chiplet (and hence the "current" row
+    # must not be accessed).
+    #
+    # As a result, `c_i` does not include the constraint of being in the memory chiplet, or in the
+    # same context and word - these must be enforced separately.
+    let f = [f0, f1, f2, f3];
+    let c = [compute_c_i(f_i, is_read', is_word_access') for f_i in f];
+
+    # Enforce that when v'[i] is not written to, then v'[i] must be 0.
+    enf v0' = 0 when c[0];
```

Comment:
The docs show the first-row constraint as: fmem_fr*ci*v'i = 0. This implementation applies constraints when c[i] = 1, but the original constraint gates this with fmem_fr flag. Should this include the fmem_fr gating condition?